crates/mdbook-mun-examples-test/src/main.rs

```diff
@@ -0,0 +1,120 @@
+use mdbook::renderer::RenderContext;
+use mdbook::BookItem;
+
+use std::io;
+
+use regex::Regex;
+
+use mun_test::{CompileAndRunTestDriver, CompileTestDriver};
+
+use mun_runtime::RuntimeBuilder;
+
+use std::panic;
+
+enum TestingMethod {
+    Skip,
+    CompileFail,
+    CompileAndRunFail,
+    Compile,
+    CompileAndRun, // default
+}
+
+fn get_testing_method(code: &str) -> TestingMethod {
+    let mut testing_method = TestingMethod::CompileAndRun;
+    for line in code.lines() {
+        if let Some(first_char) = line.chars().next() {
+            if first_char != ',' {
+                return testing_method;
+            }
+        } else {
+            return testing_method;
+        }
+        if line == ",compile_fail" {
+            match testing_method {
+                TestingMethod::Compile => testing_method = TestingMethod::CompileFail,
+                TestingMethod::CompileAndRun => testing_method = TestingMethod::CompileAndRunFail,
+                _ => (),
+            }
+        }
+        if line == ",no_run" {
+            match testing_method {
+                TestingMethod::CompileAndRun => testing_method = TestingMethod::Compile,
+                TestingMethod::CompileAndRunFail => testing_method = TestingMethod::CompileFail,
+                _ => (),
+            }
+        }
+        if line == ",skip" {
+            return TestingMethod::Skip;
+        }
+    }
+
+    testing_method
+}
+
+fn config_fn(runtime_builder: RuntimeBuilder) -> RuntimeBuilder {
+    runtime_builder
+}
+
+fn test_code(code: &str) {
+    let testing_method = get_testing_method(code);
+
+    // Removing '#' from code
+    let mut code = code.replace("\n#", "\n");
+
+    // Removing testing commands like "no_run", "compile_fail"
+    for testing_command in ["no_run", "compile_fail"].iter() {
+        code = code.replace(format!(",{}", testing_command).as_str(), "");
+    }
+
+    match testing_method {
+        TestingMethod::CompileFail => {
+            let previous_hook = panic::take_hook();
+            panic::set_hook(Box::new(|_| {}));
+
+            if panic::catch_unwind(|| CompileTestDriver::new(&code)).is_ok() {
+                panic::set_hook(previous_hook);
+                panic!("Code that should have caused the error compiled successfully ❌")
+            }
+
+            panic::set_hook(previous_hook);
+        }
+        TestingMethod::CompileAndRunFail => {
```

Comment:
`CompileAndRun` is a little trickier in the case of Mun, because you'd have to specify what function should be "run".

E.g. given a Mun file:

```mun
pub fn foo() {
  // Do something
}
```

You'd need to:
```rust
let runtime = CompileAndRunTestDriver::new(&code, config_fn).unwrap();

let _: () = invoke_fn!(runtime, "foo").unwrap();
```

Only if the `invoke_fn!`  macro fails (or not), do we want to handle the error. This might require a slightly different approach to "compile and run" for Mun.

E.g. we could opt for `no_run` by default. If we do want to run a test, we can add an annotation `run(path/to/listing.rs)` to run the Mun macro.

What do you think @baszalmstra, is this necessary for the current PR?

Reply:
I would love it if that was added!

Reply:
We could also make it only work for a specific signature? `fn main() {}`?

Reply:
I started to create keyword `fn("*")` that allows to specify what function to run in CompileAndRun mode. So `invoke_fn!` requires specify return type of specifed function. So how i can solve this? Create `match` for every mun/rust type? I mean like:
```
match function_return_type {
 "i64" => { let _: i64 = invoke_fn!(runtime, "foo").unwrap(); }
 "i32" => { let _: i32 = invoke_fn!(runtime, "foo").unwrap(); }
 "bool" => { let _: bool = invoke_fn!(runtime, "foo").unwrap(); }
...
}
```

Reply:
I think that would become very tricky. There are two options:

1) Add the `run(path/to/listing.rs)` to run the Mun macro, such that you can write whatever test code you want in plain Rust. That prevents you from needing to parse.
2) (For the time being) limit the signatures to function that don't take an argument and don't return a result. You can do so:
    a) as Bas said, by requiring a `pub fn main() { /* put test code here */ }`  in the Mun source file
    b) add a `fn("*")` macro that specifies the function name, such that you can at least change the function name
  I don't know that option b) really adds a lot of value, so maybe stick with option a for now, and add `pub fn main()` to the Mun code that you want to test?

Reply:
Added `run` keyword that's finds function with `run` name in the code and run it(`decided to take such a name for the function for clarity, but i think it's can be even more confusing, so please tell me what you think function name should be).

Reply:
I think `test` would work. You can just default to that, without anyone needing to specify it